### ISSUE TITLE
Build process has strange behavior affecting tests

### ISSUE THREAD

**Issue by soltanmm** [grpc/grpc#1057]
Replicate:
```
make CONFIG=dbg
make CONFIG=opt
```

One would expect, given we've partitioned our build targets into different folders based on configuration (e.g. `CONFIG=dbg` vs. `CONFIG=opt`), that running the command `make CONFIG=dbg` following the above would result in nothing being done. However, several files are re-built and re-linked:

```
[C]       Compiling src/core/security/credentials.c
[C]       Compiling src/core/security/json_token.c
[C]       Compiling src/core/tsi/ssl_transport_security.c
[AR]      Creating libs/dbg/libgrpc.a
[LD]      Linking libs/dbg/libgrpc.so.0.5.0.0
[LD]      Linking libs/dbg/libgrpc++.so.0.5.0.0
```

This hides #1056.

**Comment by nicolasnoble** [comment]
The rootcause for the security code to recompile in that case is because it's building OpenSSL in-place in the third_party folder. Here's the full 9 yards of what's going on:

1. You compile with CONFIG=dbg
  * You don't have OpenSSL already compiled in debug mode.
  * OpenSSL gets compiled in-place with CONFIG=dbg
  * OpenSSL's configure script touches all of its headers.
2. You compile with CONFIG=opt
  * You don't have OpenSSL already compiled in release mode.
  * OpenSSL gets compiled in-place with CONFIG=opt
  * OpenSSL's configure script touches all of its headers.
3. You compile with CONFIG=dbg
  * You already have OpenSSL compiled in debug mode - no need for recompilation.
  * credentials.c, json_token and ssl_transport_security.c all depend on OpenSSL headers
  * However, in step 2, OpenSSL's headers got touched.
  * Makefile's dependency tree notices that and considers these 3 files need to be recompiled.
  * As these files are getting recompiled, the libraries need relinking.

Long story short: if you want to avoid this, install OpenSSL 1.0.2 on your system.

**Comment by soltanmm** [comment]
Closing!
